modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -1298,6 +1319,93 @@ private CompletableFuture<Boolean> onPrimaryExpired(PrimaryReplicaEventParameter
             return falseCompletedFuture();
         }
 
+        // TODO: move to Replica https://issues.apache.org/jira/browse/IGNITE-23750
+        private void registerFailoverCallback(TablePartitionId replicationGroupId) {
+            CompletableFuture<Replica> replicaFuture = replicaManager.replica(replicationGroupId);
+
+            // TODO: https://issues.apache.org/jira/browse/IGNITE-23753
+            // For now we assume that it's a possible situation to get null future on node's stopp process, but it should be fixed with
+            // a correct and consistent replica's state.
+            if (replicaFuture == null) {
+                return;
+            }
+
+            replicaFuture.thenApply(electedPrimaryReplica -> {
+                onLeaderElectedFailoverCallback = (leaderNode, term) -> changePeersAndLearnersAsyncIfPendingExists(
+                        electedPrimaryReplica,
+                        replicationGroupId,
+                        term
+                );
+
+                return electedPrimaryReplica.raftClient().subscribeLeader(onLeaderElectedFailoverCallback);
+            }).exceptionally(e -> {
+                LOG.error("Rebalance failover subscription on elected primary replica failed [groupId=" + replicationGroupId + "].", e);
+
```

Comment:
Should we trigger failHandler in that case?

Reply:
I bet no, because the exception means one of:
1. we couldn't subscribe our client and then won't get new events with terms;
2. we couldn't get a leader, so we can't perform anything with failover then.

Reply:
My fault. By failover I mean failHandler
`failureManager.process(new FailureContext(CRITICAL_ERROR, e));`

Reply:
Added, but it's usage demands `busyLock` guarding then. Also I'm not sure: should I do retries then on L1448-1450? Failure handler looks wrong there.

Reply:
Please consider using `org.apache.ignite.internal.util.IgniteUtils#inBusyLockSafe` or `org.apache.ignite.internal.util.IgniteUtils#inBusyLock(org.apache.ignite.internal.util.IgniteSpinBusyLock, java.lang.Runnable)` or similar.

Reply:
> Also I'm not sure: should I do retries then on L1448-1450?

Just add a todo https://issues.apache.org/jira/browse/IGNITE-23633

Reply:
Done